### ISSUE TITLE
blocking thread

### ISSUE THREAD

**Issue by TrabacchinLuigi** [dokan-dev/dokan-dotnet#87]
I do think that `NativeMethods.DokanMain` locking the thread while something happens is more annoing than useful and would be better if that function would return after having the file system mounted and then implement some error handling function in the `IDokanOperations` for errors that could happen in the future while immediate errors could stay there

**Comment by Liryna** [comment]
Hi @TrabacchinLuigi ,

I got a couple of people requesting this features and I always propose to simply call DokanMain in a new thread but I am open to any proposition.

What kind of implementation for errors handling would your propose ?

**Comment by TrabacchinLuigi** [comment]
Not really sure how  i don't work with unmanaged code that much...
I would like to just have DokanMain return if everythings goes right and have a HandleDokanError that get's called when docan is going down unexpectedly ... like someone unmounted your drive, or dokany become unavailable for some reason ...
I can share how i actually handle mounting and unmounting the mirror to show why it is a little annoying, also my implementation is disposable and unmount cleanly the disk before shutting down ...

**Comment by Liryna** [comment]
I would prefer to let DokanMain C as it is :smiley: 
I was thinking to make async DokanMain by C# wrapper (Start thread, call event when return..)

I would be interested to see show this is annoying for you, yes.

**Comment by TrabacchinLuigi** [comment]
here we go: [How i would write mirror.cs and use it](https://gist.github.com/TrabacchinLuigi/da5d42a6fab5ec76ef885d904a881936)

**Comment by TrabacchinLuigi** [comment]
The annoying part is i have to wait for a semaphore (line 69, that get's green when an exception is thrown during mount or when mounted get's called)... it's not that intuitive ...
Average users would not understand how to use it...
Average user want to do programmatically call like:
```
Using(var dokan = new DocanNetImplementation()) 
{ 
    dokan.Mount("C:\myfolder", "N:\"); // expect that when this call return N:\ is available
    System.IO.CreateFile("N:\myfile.txt");
}
```

**Comment by marinkobabic** [comment]
I think @TrabacchinLuigi is right. For the moment I would keep it like it is until the following issue has been implemented. https://github.com/dokan-dev/dokany/issues/210

**Comment by Liryna** [comment]
@TrabacchinLuigi You made your point :+1: 
And @marinkobabic is right, this would be solved by issue dokan-dev/dokany#210
We would have DokanMain C return at the Mount.

It also mean that the DokanMain async will probably more be in 1.1.0 than in the next release.

**Comment by TrabacchinLuigi** [comment]
Great! keep up the good work!
i'll keep this open as a remainder!
